### PR TITLE
Add stub for lxml.etree.TreeBuilder

### PR DESCRIPTION
**Pull request by hoefling**
Signed-off-by: oleg.hoefling <oleg.hoefling@gmail.com>

### CODE REVIEW

File: lxml-stubs/etree.pyi [hunk 1]
```diff
@@ -470,3 +471,26 @@ class XPath:
         _etree_or_element: Union[_Element, _ElementTree],
         **_variables: _XPathObject
     ) -> _XPathObject: ...
+
+_ElementFactory = Callable[[Any, Dict[Any, Any]], _Element]
```
Comment (scoder) [inline lxml-stubs/etree.pyi:475]: I think we can safely assume the attribute dict to be
```suggestion
_ElementFactory = Callable[[Any, Dict[_AnyStr, _AnyStr]], _Element]
```

File: lxml-stubs/etree.pyi [hunk 2]
```diff
@@ -470,3 +471,26 @@ class XPath:
...
+_CommentFactory = Callable[[Optional[_AnyStr]], _Comment]
+_ProcessingInstructionFactory = Callable[
+    [_AnyStr, Optional[_AnyStr]], _ProcessingInstruction
+]
```
Comment (scoder) [inline lxml-stubs/etree.pyi:479]: The way these factories are used makes the argument non-optional.
```suggestion
_CommentFactory = Callable[[_AnyStr], _Comment]
_ProcessingInstructionFactory = Callable[
    [_AnyStr, _AnyStr], _ProcessingInstruction
]
```

File: lxml-stubs/etree.pyi [hunk 3]
```diff
@@ -470,3 +471,26 @@ class XPath:
...
+        pi_factory: Optional[_ProcessingInstructionFactory] = ...,
+        insert_comments: bool = ...,
+        insert_pis: bool = ...,
+    ) -> None: ...
+    def close(self) -> _Element: ...
+    def comment(self, text: _AnyStr) -> None: ...
+    def data(self, data: _AnyStr) -> None: ...
+    def end(self, tag: _AnyStr) -> None: ...
+    def pi(self, target: _AnyStr, data: Optional[_AnyStr] = ...) -> Any: ...
+    def start(self, tag: _AnyStr, attrib: _DictAnyStr) -> None: ...
```
Comment (scoder) [inline lxml-stubs/etree.pyi:496]: I don't think `_DictAnyStr` is correct here (and probably elsewhere). There is no guarantee that a unicode name will map to a unicode value and vice versa (in Py2, their type is independent). In Py3, it's always unicode text anyway.

**Review by scoder — COMMENTED** [review]
Thanks. See my comments below.

File: lxml-stubs/etree.pyi (hunk 3 above)
Comment (hoefling) [reply lxml-stubs/etree.pyi]: @scoder I agree, thought about that too - however, I have decided not to go against the current typing of the `ParserTarget` protocol. Now that you have mentioned it, I have adjusted the typing of `ParserTarget.start` as well.

Comment (hoefling) [reply lxml-stubs/etree.pyi]: I've also added a test that should fail if `TreeBuilder` becomes incompatbile to the `ParserTarget` spec.

**Comment by scoder** [comment]
Thanks

**Pull request merged**